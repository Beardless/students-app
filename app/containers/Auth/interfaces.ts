import { ModuleInterface } from "../SingleCourse/interfaces";

export interface AuthorInterface {
  bio: string;
  href: {
    picture: string;
  };
  name: string;
  picture: string;
}

interface AuthFields {
  type: string;
  label: string;
  value: string;
  disabled: boolean;
  error: string;
}

export interface CourseInterface {
  id?: number;
  title: string;
  description: string;
  cover_image_name: string;
  created: string;
  updated: string;
  href: {
    cover_image_url: string;
  };
  modules?: ModuleInterface[];
  authors?: AuthorInterface[];
}

export interface UserInterface {
  email: string;
  created: string;
  first_name: string;
  last_name: string;
  id: string;
  updated: string;
  roles: string[];
  courses: CourseInterface[];
}

export interface AuthFieldsInterface extends Array<AuthFields> {}
