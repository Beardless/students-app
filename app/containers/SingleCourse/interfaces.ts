export interface ModuleInterface {
  id: string;
  title: string;
  description: string;
  position: number;
  created: string;
  updated: string;
  course: {
    id: number;
    title: string;
  };
  lessons: { id: string; title: string }[];
}
