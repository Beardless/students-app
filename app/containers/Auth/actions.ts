import {
  AUTH_ACTION,
  AUTH_ACTION_ERROR,
  AUTH_ACTION_SUCCESS,
  userLoginInterface
} from "./constants";

export function authAction(
  payload: userLoginInterface
): { type: string; payload: userLoginInterface } {
  return {
    type: AUTH_ACTION,
    payload
  };
}

export function authActionSuccess(user: any): { type: string; user: any } {
  return {
    type: AUTH_ACTION_SUCCESS,
    user
  };
}

export function authActionError(
  error: string
): { type: string; error: string } {
  return {
    type: AUTH_ACTION_ERROR,
    error
  };
}
