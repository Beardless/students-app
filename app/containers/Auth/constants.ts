export const AUTH_ACTION: string = "courses/Auth/AUTH_ACTION";
export const AUTH_ACTION_SUCCESS: string = "courses/Auth/AUTH_ACTION_SUCCESS";
export const AUTH_ACTION_ERROR: string = "courses/Auth/AUTH_ACTION_ERROR";
export const USER_ACTION: string = "courses/Auth/USER_ACTION";

export interface userLoginInterface {
  username: string;
  password: string;
}
