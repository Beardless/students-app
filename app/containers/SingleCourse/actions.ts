import {
  COURSE_ACTION,
  COURSE_ACTION_SUCCESS,
  COURSE_ACTION_ERROR
} from "./constants";
import { ModuleInterface } from "./interfaces";

export function courseAction(
  id: string
): { type: string; payload: { id: string } } {
  return {
    type: COURSE_ACTION,
    payload: { id }
  };
}

export function courseActionSuccess(
  modules: [ModuleInterface]
): { type: string; modules: [ModuleInterface] } {
  return {
    type: COURSE_ACTION_SUCCESS,
    modules
  };
}

export function courseActionError(
  error: string
): { type: string; error: string } {
  return {
    type: COURSE_ACTION_ERROR,
    error
  };
}
