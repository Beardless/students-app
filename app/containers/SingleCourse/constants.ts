export const COURSE_ACTION: string = "courses/SingleCourse/COURSE_ACTION";
export const COURSE_ACTION_SUCCESS: string =
  "courses/SingleCourse/COURSE_ACTION_SUCCESS";
export const COURSE_ACTION_ERROR: string =
  "courses/SingleCourse/COURSE_ACTION_ERROR";
