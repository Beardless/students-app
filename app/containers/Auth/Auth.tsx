import * as React from "react";
import i18n from "i18next";
import { Redirect } from "react-router-dom";
import { userLoginInterface } from "./constants";
import Login from "../../components/Login";
import "./style.scss";
import { AuthFieldsInterface } from "./interfaces";
import { getValueFromFields } from "./auth-logic";

interface AuthProps {
  onLogin: (payload: userLoginInterface) => void;
  loading: boolean;
  error: string;
  currentUser: any;
}

interface AuthState {
  fields: AuthFieldsInterface;
}

class Auth extends React.Component<AuthProps, AuthState> {
  constructor(props) {
    super(props);
    this.state = {
      fields: [
        {
          type: "login",
          label: i18n.t("Login"),
          value: "",
          disabled: false,
          error: null
        },
        {
          type: "password",
          label: i18n.t("Password"),
          value: "",
          disabled: false,
          error: null
        }
      ]
    };
  }

  handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { fields } = this.state;
    const payload = {
      username: getValueFromFields(fields, "login"),
      password: getValueFromFields(fields, "password")
    };

    this.props.onLogin(payload);
  };

  handleUserInput = (fieldName: string, value: string) => {
    const { fields } = this.state;
    const updatedFields = fields.map(field => {
      if (field.type === fieldName) {
        field.value = value;
        return field;
      }
      return field;
    });

    this.setState({ fields: updatedFields });
  };

  render() {
    const { fields } = this.state;
    const { loading, error, currentUser } = this.props;

    if (currentUser) {
      return <Redirect to={"/"} />;
    }

    return (
      <div className={"container"}>
        <Login
          fields={fields}
          isLoading={loading}
          error={error}
          onUserInput={this.handleUserInput}
          onSubmit={this.handleSubmit}
        />
      </div>
    );
  }
}

export default Auth;
