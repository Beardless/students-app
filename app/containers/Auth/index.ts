import { connect } from "react-redux";
import { compose } from "redux";
import { authAction } from "./actions";
import { createStructuredSelector } from "reselect";
import { userLoginInterface } from "./constants";
import Auth from "./Auth";
import {
  makeSelectAuthError,
  makeSelectAuthLoading,
  makeSelectUser
} from "./selectors";

const mapDispatchToProps = dispatch => ({
  onLogin: (payload: userLoginInterface) => dispatch(authAction(payload))
});

const mapStateToProps = createStructuredSelector({
  loading: makeSelectAuthLoading(),
  error: makeSelectAuthError(),
  currentUser: makeSelectUser()
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);
// @ts-ignore
export default compose(withConnect)(Auth);
