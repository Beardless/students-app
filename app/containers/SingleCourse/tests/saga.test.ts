import getSingleCourse, { getCourse } from "../saga";
import * as course from "./mocks/course.json";
import * as user from "./../../Auth/tests/mocks/user.json";
import { put, takeEvery } from "redux-saga/effects";
import { courseActionError, courseActionSuccess } from "../actions";
import { COURSE_ACTION } from "../constants";
import { CourseInterface } from "../../Auth/interfaces";

const coursePayload = { payload: { id: "1" } };

describe("getCourse Saga", () => {
  let getCourseGenerator;

  beforeEach(() => {
    getCourseGenerator = getCourse(coursePayload);
    const callDescriptor = getCourseGenerator.next(coursePayload).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("should dispatch the courseActionSuccess if it requests the data successfully", () => {
    const response: { data: CourseInterface } = { data: course };
    getCourseGenerator.next({ data: response.data.modules });
    const putDescriptor = getCourseGenerator.next({ courses: user.courses })
      .value;
    expect(putDescriptor).toEqual(put(courseActionSuccess(response.data)));
  });

  it("should call the authActionError if the response errors", () => {
    const err = { response: { data: { message: "Test error" } } };
    const putDescriptor = getCourseGenerator.throw(err).value;
    expect(putDescriptor).toEqual(put(courseActionError("Test error")));
  });
});

describe("getSingleCourse Saga", () => {
  const getSingleCourseSaga = getSingleCourse();

  it("should start task to watch for COURSE_ACTION action", () => {
    const takeEveryDescriptor = getSingleCourseSaga.next().value;
    // @ts-ignore
    expect(takeEveryDescriptor).toEqual(takeEvery(COURSE_ACTION, getCourse));
  });
});
