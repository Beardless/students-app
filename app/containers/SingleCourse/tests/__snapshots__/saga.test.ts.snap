// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCourse Saga should call the authActionError if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "url": "/courses/1/modules",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getCourse Saga should dispatch the courseActionSuccess if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "url": "/courses/1/modules",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
